### ISSUE TITLE
Plots.jl optional backend files are not tracked

### ISSUE THREAD

**Issue by isentropic** [timholy/Revise.jl#636]
Moving https://github.com/JuliaPlots/Plots.jl/issues/3345 for more visibility. 
I confirmed a few things with other maintainers of Plots:
1. Optional backend files are not tracked: "src/pyplot, pgfplotsx.jl" and other backends
2. Standard backend files "gr.jl, plotly.jl" work as their are known to be included compile time I think.

Some clues are here https://github.com/JuliaPlots/Plots.jl/issues/3345#issuecomment-799144457

I just wan to say that this hinders the development a lot as Plots is notorious for long inclusion times.

**Comment by isentropic** [comment]
So I had to learn a few things about requires.jl and what it is; however even if I:
```
julia> using Revise

julia> import PyPlot
 
julia> using Plots
```
(or any other backend file) Revise doesn't work with /src/backends/pyplot.jl code base

**Comment by isentropic** [comment]
I even discovered https://github.com/timholy/Revise.jl/issues/316 (evidence that it has worked in the past) right now it does not. I'm fairly certain that there haven't been any changes related to the optional backend loading since that time.

**Comment by aloy1496** [comment]
Where are you

**Comment by timholy** [comment]
> Where are you

Next to the kitchen, why do you ask? Seriously, you're not paying me anything and I am not your servant. Some of us have grants to write so we can pay the people who work in our group.

**Comment by t-bltg** [comment]
When `Revise` parses `Plots`, it uses the `Plots` precompiled file:
https://github.com/timholy/Revise.jl/blob/165433a5e4fc5cdcee1d2ce20843ad476f1c5a83/src/loading.jl#L25

When precompilation occurs, only these backends are loaded: https://github.com/JuliaPlots/Plots.jl/blob/cebcdfb3b51966f256dbf68e9e1639f3b50523b7/src/Plots.jl#L216.

So when you load an optional backend e.g. `Gaston` then `src/backends/gaston.jl` is absent from the list of includes: https://github.com/timholy/Revise.jl/blob/165433a5e4fc5cdcee1d2ce20843ad476f1c5a83/src/loading.jl#L31.

On solution to this, when developing `Plots` is to disable precompilation of `Plots`: by adding `__precompile__(false)` at the head of module Plots [here](https://github.com/JuliaPlots/Plots.jl/blob/cebcdfb3b51966f256dbf68e9e1639f3b50523b7/src/Plots.jl#L2), you force `Revise` to watch files registered in `included_files`: https://github.com/timholy/Revise.jl/blob/165433a5e4fc5cdcee1d2ce20843ad476f1c5a83/src/loading.jl#L51.

So now that I added `__precompile__(false)` in Plots, starting julia and entering the following commands show that the optional backend `src/backends/gaston.jl` is correctly tracked by `Revise`
```julia
julia> using Revise
julia> import Gaston
julia> using  Plots
[ Info: Precompiling Plots [91a5bcdd-55d7-5caf-9e0b-520d859cae80]
[ Info: Skipping precompilation since __precompile__(false). Importing Plots [91a5bcdd-55d7-5caf-9e0b-520d859cae80].
julia> "src/backends/gaston.jl" ∈ Revise.CodeTracking._pkgfiles[Base.PkgId(Base.UUID("91a5bcdd-55d7-5caf-9e0b-520d859cae80"), "Plots")].files
true
```

I can now develop `src/backends/gaston.jl` without having to restart julia.

**Comment by timholy** [comment]
For anyone who wants to help get this working, maybe try to figure out why the [LikePlots tests](https://github.com/timholy/Revise.jl/blob/165433a5e4fc5cdcee1d2ce20843ad476f1c5a83/test/runtests.jl#L2697-L2740) are not representative of the way modern plots actually works (I haven't looked yet).

**Comment by isentropic** [comment]
I see with Requires jl, the optional backends are not tracked due tue precompiles. Is this a change in Julia and precompilation ? Maybe for `dev`ed packages precomplilation and this issue could somehow be hacked?

**Comment by t-bltg** [comment]
> For anyone who wants to help get this working, maybe try to figure out why the [LikePlots tests](https://github.com/timholy/Revise.jl/blob/165433a5e4fc5cdcee1d2ce20843ad476f1c5a83/test/runtests.jl#L2697-L2740) are not representative of the way modern plots actually works (I haven't looked yet).

Maybe because the test [explicitly tracks](https://github.com/timholy/Revise.jl/blob/165433a5e4fc5cdcee1d2ce20843ad476f1c5a83/test/runtests.jl#L2719) the backend files ? Also, there do not seem to be any precompilation cache in this tests, does it ?
Yes, it's a good idea to revise the `LikePlots` test, unfortunately for the moment I'm a bit busy and this is low on my todo list.
But maybe @isentropic can help here, it's a good way to dig deeper into `Revise` and the way julia loads modules.

> I see with Requires jl, the optional backends are not tracked due tue precompiles. Is this a change in Julia and precompilation ? Maybe for deved packages precomplilation and this issue could somehow be hacked?

Well, I jumped in the julia world too recently to be able to say if that is a regression or what was the behavior before ...
I'm a bit confused here: I already provide a hack [here](https://github.com/timholy/Revise.jl/issues/636#issuecomment-888372797): since you're developing `Plots`, **add `__precompile__(false)` at the top level of the `Plots` module**.
If you want to push it further you can always define an env variable for triggering `dev` mode like so:
```julia
module Plots
  @static if get(ENV, "PLOTS_DEV", "false") == "true"
      __precompile__(false)
  end
  [...]
```
, and start the `REPL` like this: `$ PLOTS_DEV=true julia`.

But I don't think that this should go permanently into `Plots.jl`.

**Comment by timholy** [comment]
It turns out to be a one-line fix (#636). And my tests had an extra line or I would have caught this.

If I were to guess, this was a victim of #478 (deletion of the `Base.convert(::Type{RelocatableExpr}, ex::Expr) = RelocatableExpr(ex)` method), so it's strange this didn't come up before.
